File: _Season15/01 - Not Quite Lisp/Python/aoc2015_01_2.py
"""
aoc2015_01_1.py
----------------
Author: Nida Anis
Date created: 20/02/2025
----------------
Description:
- Advent of Code 2015 Day 1: Not Quite Lisp
- Solution to Part 2 (Part 1 in aoc2015_01_1.py)
"""

from aoc2015_01_1 import get_map_from_file

def get_first_basement_pos(map):
    """Gets the first position of the character that causes Santa to enter the basement."""
    position = 0
    floor_num = 0
    basement_positions = []

    for char in map:
        position += 1

        if char == "(":
            floor_num += 1
        
        elif char == ")":
            floor_num -= 1
        
        if floor_num == -1:
            basement_positions.append(position)
    
    return basement_positions

def main():
    map = get_map_from_file()
    basement_positions = get_first_basement_pos(map)
    print(f"The position of the first character that causes Santa to enter the basement is {basement_positions[0]}.")

if __name__ == "__main__":
    main()

File: _Season15/01 - Not Quite Lisp/C#/NotQuiteLisp/aoc2015_01_2.cs
/*
aoc2015_01_2.cs
----------------
Author: Nida Anis
Date: 31/12/2024
----------------
Description:
- Advent of Code 2015 Day 1: Not Quite Lisp
- Solution to Part 2 (Part 1 in aoc2015_01_1.cs)
*/

class Part2
{

    /// <summary>
    /// Gets the position of the first character that causes Santa to enter the basement.
    /// </summary>
    /// <param name="charArray"></param>
    /// <returns>The position of the first character that causes Santa to enter the basement.</returns>
    public static int GetFirstBasementPos(char[] charArray)
    {
        int pos = 0;
        int floorNumber = 0;

        for (int i = 0; i < charArray.Length; i++)
        {
            pos += 1;

            if (charArray[i] == '(')
            {
                floorNumber += 1;
            }
            else if (charArray[i] == ')')
            {
                floorNumber -= 1;
            }

            if (floorNumber == -1)
            {
                return pos;
            }
        }

        return -1;
    }
}